### PR TITLE
Replace ref counting with shared ptr in ServiceReferenceBase class family

### PR DESCRIPTION
**Pull request by karthikreddy09**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -257,6 +257,17 @@ endif()
...
+if(NOT APPLE AND US_COMPILER_CLANG)
+  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -stdlib=libc++ -lc++abi")
+  set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} -stdlib=libc++ -lc++abi")
+  set(CMAKE_MODULE_LINKER_FLAGS "${CMAKE_MODULE_LINKER_FLAGS} -stdlib=libc++ -lc++abi")
+endif()
+
+if(US_COMPILER_APPLE_CLANG OR US_COMPILER_CLANG)
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -stdlib=libc++")
+endif()
+
```
[the remark this answers is not in the document]
Comment (karthikreddy09) [reply CMakeLists.txt]: Unfortunately, they are tied together. atomic_load and atomic_store operations for shared_ptr are not supported in GCC4.9 even with C++11. Our clang builds are picking up GCC 4.9 headers/libstdc++ by default. With the switch to libc++, I got past those build errors.

File: CMakeLists.txt [hunk 2]
```diff
@@ -297,55 +308,54 @@ configure_file(
...
-    set(BENCHMARK_ENABLE_INSTALL OFF)
-    set(BENCHMARK_ENABLE_TESTING OFF CACHE BOOL "" FORCE)
-    add_subdirectory(third_party/benchmark)
+  set(BENCHMARK_ENABLE_INSTALL OFF)
+  set(BENCHMARK_ENABLE_TESTING OFF CACHE BOOL "" FORCE)
+  add_subdirectory(third_party/benchmark)
 
-    set(BUILD_SHARED_LIBS ${_us_build_shared} CACHE BOOL "Build shared libraries" FORCE)
-  endif()
+  set(BUILD_SHARED_LIBS ${US_BUILD_SHARED_LIBS} CACHE BOOL "" FORCE)
```
[the remark this answers is not in the document]
Comment (karthikreddy09) [reply CMakeLists.txt]: Thanks for spotting this. I will fix it.

File: CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (saschazelzer) [reply CMakeLists.txt]: I see, that wasn't obvious to me thanks. Maybe add a note in the CMake files as well?

File: framework/src/service/ServiceReferenceBase.cpp [hunk 3]
```diff
@@ -38,23 +38,32 @@ ServiceReferenceBase::ServiceReferenceBase()
...
+  : d(ref.GetPrivate())
+{}
 
 ServiceReferenceBase::ServiceReferenceBase(ServiceRegistrationBasePrivate* reg)
   : d(new ServiceReferenceBasePrivate(reg))
 {}
 
 void ServiceReferenceBase::SetInterfaceId(const std::string& interfaceId)
 {
-  if (d.load()->ref > 1) {
```
[the remark this answers is not in the document]
Comment (karthikreddy09) [reply framework/src/service/ServiceReferenceBase.cpp]: Do you have any suggestions for this. I don't think we can use the shared_ptr::use_count to replicate previous logic. 

Constructing ServiceReferenceBasePrivate objects is not costly considering the only data it has is a pointer (ServiceRegistrationBasePrivate*) and a std::string. The constructor used here copies the ServiceRegistrationBasePrivate raw pointer so we are not creating a new ServiceRegistration object. Do we need to optimize this given the possibility of a race?

File: third_party/benchmark/src/benchmark.cc [hunk 4]
```diff
@@ -153,11 +154,13 @@ State::State(size_t max_iters, const std::vector<int64_t>& ranges, int thread_i,
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Winvalid-offsetof"
 #endif
+#if defined(__APPLE__) || !defined(__clang__) // remove the following block for clang on linux to avoid compile error
```
Comment (jeffdiclemente) [inline third_party/benchmark/src/benchmark.cc]: a patch file should be added to the third_party directory

File: framework/src/service/ServiceReferenceBase.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (saschazelzer) [reply framework/src/service/ServiceReferenceBase.cpp]: Sorry for taking so long. My concern is with the memory allocation in `make_shared`, which is new compared to the optimized path for `ref == 1` before.

Initially, I couldn't think of a good way to handle this. But now when I look at the call sites of  `SetInterfaceId` (there are only two), we only call it on newly constructed `ServiceReference` instances (and the `SetInterfaceId` is not public). So they cannot be possibly shared and hence there is no need for this complication at all. We can just simply do `GetPrivate()->interfaceId = interface` and be done with it. If you agree, please add a comment to the `SetInterfaceId` method declaration the it must no be used on shared instances.

**Pull request closed without merging**